Remove debug output - minimal version

Finii opened the pull request:
This is in some ways less and more than
* #13

It does
* It removes the debug output from `stdout`
* But instead of forwarding the message to be `throw`n later it is just completely dropped
* But is does implement an operator bool to avoid all the tests for `nullptr`
* Adding the extra throwing behavior on top of this change (like PR 13 did) is an easy next step, if we want to go that way
* So this improves the code but does not add more functionality

Apart from that, and ahead of #13 it renames `LibGitPointer` to `LibGitObject` and later does not directly use that anymore but uses 'named types' (aliases) for the various objects.

I believe `Pointer` is a misnomer. Of course technically it just holds a pointer, and it just calls 'free()' in the appropriate moment. But that is exactly the definiting of 'owing' the structure. So from a C++ point of view this 'wrapper object' **is** the object.

This is why I wanted to have (for example) a `LibGitCommit` which looks like a C++ object and behaves like a C++ object albeit under the hood it is a c pointer. Well, it is, but that does not need to show by the name.

Which does look more like 'libgit2 for cpp'? :

```c++
// implemented in this PR
LibGitCommit GitRepository::get_commit(unsigned int count);

 // previous naming
LibGitPointer<git_commit> GitRepository::get_commit(unsigned int count);
```

Fixes: #5
Fixes: #15

Finii committed:
wrapper_functions: Remove explicit object creation

Signed-off-by: Fini Jastrow <ulf.fini.jastrow@desy.de>

Finii committed:
wrapper_functions: Improve docs

Signed-off-by: Fini Jastrow <ulf.fini.jastrow@desy.de>

Finii committed:
Replace LibGitPointer with unique_ptr

[why]
Because it is a unique_ptr and nothing else, they say.

Nobody cares if they all come from the same template, from a user
perspective they are all completely different beasts.

[how]
Just drop all hand written stuff and use STLs solution.

Give em all individual names and use them thus.
Makes the code look simpler.

Signed-off-by: Fini Jastrow <ulf.fini.jastrow@desy.de>

Finii committed:
wrapper_functions: Drop debug messages

But keep them as comments.

Signed-off-by: Fini Jastrow <ulf.fini.jastrow@desy.de>

Finii committed:
GitRepository: Remove nonexisting function declaration

[why]
Obviously this has been forgotten when the constructors have been unified.

Signed-off-by: Fini Jastrow <ulf.fini.jastrow@desy.de>

Finii commented:
I'm not sure about the last commit

629cea2a `GitRepository: Make more members const`

I would do that, though, but if you are uncomfortable with `const_cast` we can drop that commit.

And when writing this, I can hardly follow why people want _const west_... For me the _east const_ keeps the two things together we also use when speaking. "That object is handed over by const-ref", then write that?

```c++
// how I finally wrote it in the PR
RepoState collect_status(const LibGitStatusList& status) const;

// what I instrinctically write first and needed to 'correct'
RepoState collect_status(LibGitStatusList const& status) const;
```

_Just trying to hand out east-const-wristbands_

![signal-2023-12-11-091402](https://github.com/taskolib/libgit4cpp/assets/16012374/e7cc8001-1cae-4696-8c75-695b3285c853)

alt-graph commented on include/libgit4cpp/GitRepository.h:
```diff
@@ -32,7 +32,7 @@
 #include <gul14/escape.h>
 
-#include "libgit4cpp/LibGitPointer.h"
+#include "libgit4cpp/LibGitObject.h"
```
I am not entirely sure I follow your arguments for the renaming. In essence, a `LibGitPointer` is a hand-rolled `unique_ptr` for Git objects. Its semantics seem more pointer-like than object-like to me: There is a test for null, and no member functions of the Git object can be called via `.`.

This is just an observation, I can live with both names.

Finii replied:
Lol, yes. Of course the unique-ptr is generic and thus there can be no better name.
This is not generic at all but holds objects we get from specifically libgit2.

And additionally it will probably get more information, like the failure message.
It does not need to `throw` like in the `feature/remove-debug-output` branch, but we can transport it within and put the work to throw appropriate objects on the user's shoulders.
At least I hoped we we add the throwing thing, but if not we can at least add the non-throwing messages.

> Its semantics seem more pointer-like than object-like to me: There is a test for null, and no member functions of the Git object can be called via

I can not see any test for null. And the whole api is (at least intended to be) alike expected or optional.
Whatever.

The main point is to have aliases for the things, to make the code readable for users.
I always want to have it easy for the users and do not care how it is handled internally, it should show nicely to the outside and be somehow logical for people who never had and never will be looking what is inside.

Assume you are new to some accounting system. You can query all the past transactions. How would you expect the API to look like? Get a `we_have_obscure_pointers<of_more_obscure_types>` or `Transaction` ?

alt-graph replied:
Don't get me wrong, I like the aliases. Still, `LibGitPtr` is just a `unique_ptr` to me, and therefore I find that name more logical than `LibGitObject`. But as I wrote, I can live with both.

alt-graph commented on src/GitRepository.cc:
```diff
@@ -226,18 +226,18 @@ LibGitCommit GitRepository::get_commit(unsigned int count)
 
     // find commit object by commit ID
-    error = git_commit_lookup(&commit, repo_.get(), &oid_parent_commit);
+    error = git_commit_lookup(&commit, const_cast<git_repository*>(repo_.get()), &oid_parent_commit);
```
I am not sure these `const_cast`s are worth doing just for the sake of making some `GitRepository` member functions `const`. After all, they might invoke undefined behavior – we simply don't know what that ugly libgit2 API is doing under the hood.
> $5.2.11/7 - "[Note: Depending on the type of the object, a write operation through the pointer, lvalue or pointer to data member resulting from a const_cast that casts away a const-qualifier68) may produce undefined behavior (7.1.5.1). ]"

Finii replied:
>  they might invoke undefined behavior

Yes, theoretically.

Whatever, removing them.

Finii commented:
Please review again.

You said / noted that `LibGitPointer` was inteded as a `std::unique_ptr` equivalent.

Then, why a special class AT ALL? Just use `unique_ptr` if you want one?

Drops a lot of hand written code, maybe this is what is really wanted?

Will revert or keep, whatever you think is intended.

Finii commented:
> Don't get me wrong, I like the aliases. Still, LibGitPtr is just a unique_ptr to me, and therefore I find that name more logical than LibGitObject. But as I wrote, I can live with both.

Ah yes. Its a `unique_ptr` for you -> then use a `unique_ptr` 😄 

But maybe the header file needs to be renamed. Any ideas?

alt-graph commented:
> > Don't get me wrong, I like the aliases. Still, LibGitPtr is just a unique_ptr to me, and therefore I find that name more logical than LibGitObject. But as I wrote, I can live with both.
> 
> Ah yes. Its a `unique_ptr` for you -> then use a `unique_ptr` 😄

Nice! I did not know that could be done.

> But maybe the header file needs to be renamed. Any ideas?

Something vague like "aliases.h" or "types.h" maybe?

Finii commented:
* Resolving the conflicts - force push imminent
* Renaming header file
* As this seems what we want: squashing some commits - force push imminent

Finii committed:
wrapper_functions: Fix and remove unused function

[why]
The function has a memory leak.
The function does not work (segfaults).
It is not used at the moment.

[how]
Fix and comment out as we might need it for ideas later if remote stuff
is implemented (again).

Signed-off-by: Fini Jastrow <ulf.fini.jastrow@desy.de>

Finii commented:
Now doing step 3, hold on

Finii commented:
Squashed a lot of commits but keeping some.
I guess this is about the gist of what this does.

Ready to merge from my side ⛱
